Source/DataSources/RelativeToTerrainHeightProperty.js

```diff
@@ -0,0 +1,201 @@
+        //>>includeStart('debug', pragmas.debug);
+        Check.defined('terrainProvider', terrainProvider);
+        //>>includeEnd('debug');
+
+        this._positions = undefined;
+        this._subscription = undefined;
+        this._heightRelativeToTerrain = undefined;
+        this._heightRelativeToTerrainSubscription = undefined;
+        this._definitionChanged = new Event();
+
+        this._terrainProvider = terrainProvider;
+        this._cartographicPosition = new Cartographic();
+        this._terrainHeight = 0;
+
+        this.positions = positions;
+        this.heightRelativeToTerrain = heightRelativeToTerrain;
+    }
+
+    defineProperties(RelativeToTerrainHeightProperty.prototype, {
+        /**
+         * Gets a value indicating if this property is constant.
+         * @memberof RelativeToTerrainHeightProperty.prototype
+         *
+         * @type {Boolean}
+         * @readonly
+         */
+        isConstant : {
+            get : function() {
+                return Property.isConstant(this._positions);
+            }
+        },
+        /**
+         * Gets the event that is raised whenever the definition of this property changes.
+         * @memberof RelativeToTerrainHeightProperty.prototype
+         *
+         * @type {Event}
+         * @readonly
+         */
+        definitionChanged : {
+            get : function() {
+                return this._definitionChanged;
+            }
+        },
+        /**
+         * Gets or sets the positions property used to compute the value.
+         * @memberof RelativeToTerrainHeightProperty.prototype
+         *
+         * @type {Property}
+         */
+        positions : {
+            get : function() {
+                return this._positions;
+            },
+            set : function(value) {
+                var oldValue = this._positions;
+                if (oldValue !== value) {
+                    if (defined(oldValue)) {
+                        this._subscription();
+                    }
+
+                    this._positions = value;
+
+                    if (defined(value)) {
+                        if (!value.isConstant) {
+                            throw new RuntimeError('positions must be a constant property');
+                        }
+                        this._subscription = value._definitionChanged.addEventListener(function() {
+                            this._fetchTerrainHeight();
+                            this._definitionChanged.raiseEvent(this);
+                        }, this);
+                    }
+
+                    this._fetchTerrainHeight();
+                    this._definitionChanged.raiseEvent(this);
+                }
+            }
+        },
+        heightRelativeToTerrain : createPropertyDescriptor('heightRelativeToTerrain')
+    });
+
+    var centroidScratch = new Cartesian3();
+    function computeCentroid(positions) {
```

Comment:
I know we discussed this offline, but if I were doing this I would make a `CentroidProperty` that computes this value and then `RelativeToTerrainHeightProperty` can just always take a PositionProperty.